### PR TITLE
Initial ellipsis support when slicing Tensors

### PR DESCRIPTION
**Pull request by fikisipi**
Assuming a 4D tensor in pytorch, `some_tensor[..., 0]` is equivalent to `some_tensor[:, :, :, 0]`
 
* [x] Pylint passes
* [x] Tests pass, torch results match tinygrad when adding ellipsis 
  * at start
  * at middle
  * at end
* [x] Check how `None` interacts with `...` 
* [x] Check out the torch / numpy spec for other edge cases

I know we have a LoC budget and you dislike slices, so feel free to close if not important enough.

### CODE REVIEW

**Comment by geohot** [comment]
If Torch/Numpy support this, it's important we do too.

**Comment by Joqsan** [comment]
Check out https://github.com/geohot/tinygrad/pull/777. We are working on supporting 0dim tensor, which include a refactor of `__getitem__`.

**Comment by fikisipi** [comment]
> Check out #777. We are working on supporting 0dim tensors, which include a refactor of `__getitem__`.

@Joqsan Did a merge at `fikisipi:slice-ellipsis2` and the ops tests pass. Your branch looks good to me but I haven't ran the ONNX tests - what else do we need to ship it?

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         8               0
------------------------------------------------------------
lines added in the tinygrad folder: 8
```

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         8               0
------------------------------------------------------------
lines added in the tinygrad folder: 8
```

**Comment by Joqsan** [comment]
> > Check out #777. We are working on supporting 0dim tensors, which include a refactor of `__getitem__`.
> 
> @Joqsan
> 
> Did a merge of this + #777 at `fikisipi:slice-ellipsis2` and the ops tests pass.
> 
> Your branch looks really good to me (i like the friendly errors as well) but I haven't ran the ONNX tests - what else do we need to ship it? Do you need any help? Personally, I think your PR should be merged first and then we can experiment with ellipses.

@fikisipi . Sorry the delay. 

That PR was merged yesterday.

I think it can be helpful to understand what's going here in these lines:

https://github.com/geohot/tinygrad/blob/4d28d5568396c1fbae6f6ee20998396bc54b30cd/tinygrad/tensor.py#L265-L268

Suppose you have a tensor `x` with `len(x.shape) = 4`. If you do slice `x[1:4]` you get a slice (1, 4) along the 0 axes. Line 267 "completes" the slice by adding `slice(None, None)` for the dimensions you didn't specify (1, 2, 3 axes). It basically does `x[1:4, :, :, :]`.

So it always completes from the right.

With ellipsis it would be the same thing (when you do x[1:4, ...], x[1:4, 3, ...]) but it also needs the ability to complete slices from the left, since, `x[..., 3]`, `x[..., 1:4, 3]` are also possible.

**Comment by geohot** [comment]
Still working on this. I like this feature.

**Comment by fikisipi** [comment]
> Still working on this? I like this feature.

@geohot Now that #777 is merged and refactors slices, I'm in a position where I can finish this PR. Likely in 7-8 hours.

**Comment by fikisipi** [comment]
* [x] `test_ops.py` passes with ellipses at start, beginning and end positions (with and without None singletons)
* [x] onnx tests aren't affected
* [x] pylint passes

## On multiple ellipses

Per the [numpy spec](https://numpy.org/doc/1.18/reference/arrays.indexing.html), multiple ellipses are not allowed:

> Ellipsis expands to the number of : objects needed for the selection tuple to index all dimensions.

Adding multiple ellipses will produce this:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: an index can only have a single ellipsis ('...')
```

In PyTorch it **is allowed and ignored** - `tensor[..., 0, ...]` will work but a current PR is planning to emit the same error: https://github.com/pytorch/pytorch/issues/59787

I follow the numpy spec and emit an `IndexError`.

## Interaction between `None` singletons and ellipsis expansion

The ellipsis will attempt to fill in *the old shape* first, and then the `None`s will alter the shape with singleton dimensions accordingly.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -263,7 +263,15 @@ def normalize_int(e, i, dim_sz):
...
-    orig_slices = list(val) + [slice(None)] * (len(self.shape) - num_slices)
+    orig_slices = list(val)
+    ellipses_found = [i for i, v in enumerate(val) if v is Ellipsis]
+    if len(ellipses_found) > 0:
+      if len(ellipses_found) != 1:
+        raise IndexError("an index can only have a single ellipsis ('...')")
+      ellipsis_idx = ellipses_found[0]
+      orig_slices[ellipsis_idx:ellipsis_idx+1] = [slice(None)] * (len(self.shape) - num_slices)
+    else:
+      orig_slices+= [slice(None)] * (len(self.shape) - num_slices)
```
Comment (geohot) [inline tinygrad/tensor.py:274]: Formatting

**Comment by geohot** [comment]
Will merge if tests pass

**Comment by fikisipi** [comment]
@geohot passes
